### PR TITLE
Add Developer guide for KFServing

### PR DESCRIPTION
**Pull request by yuzisun**
- Add perquisite installation guide
- Add test/deploy iteration guide
- Note that `make deploy-test` is introduced in https://github.com/kubeflow/kfserving/pull/65

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/67)

### CODE REVIEW

File: docs/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -1 +1,173 @@
...
+1. [`dep`](https://github.com/golang/dep): For managing external Go
+   dependencies.
+1. [`ko`](https://github.com/google/ko):
+   For development.
+1. [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/): For
+   managing development environments.
+
+### Install Knative on a Kubernetes cluster
+
+1. [Set up a kubernetes cluster and install Knative](https://knative.dev/docs/install/)
```
Comment (rakelkar) [inline docs/DEVELOPER_GUIDE.md]: possibly clarify that only knative-serving is required and knative-monitoring is good to have (and that others are currently not required?)

Comment (rakelkar) [reply docs/DEVELOPER_GUIDE.md]: I mean that you dont need to install the rest..

File: docs/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -1 +1,173 @@
...
+
+- **Note**: Set up a docker repository for pushing images. You can use any container image registry by adjusting 
+the authentication methods and repository paths mentioned in the sections below.
+   - [Google Container Registry quickstart](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+   - [Docker Hub quickstart](https://docs.docker.com/docker-hub/)
+- **Note**: if you are using docker hub to store your images your
+  `KO_DOCKER_REPO` variable should be `docker.io/<username>`.
+- **Note**: Currently Docker Hub doesn't let you create subdirs under your
+  username.
+
```
Comment (rakelkar) [inline docs/DEVELOPER_GUIDE.md]:
```suggestion
- [Azure Container Registry quickstart] (https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-portal)

```

File: docs/DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -1 +1,173 @@
...
+1. [`go`](https://golang.org/doc/install): KFServing controller is written in Go
+1. [`git`](https://help.github.com/articles/set-up-git/): For source control
+1. [`dep`](https://github.com/golang/dep): For managing external Go
+   dependencies.
+1. [`ko`](https://github.com/google/ko):
+   For development.
+1. [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/): For
+   managing development environments.
+
```
Comment (rakelkar) [inline docs/DEVELOPER_GUIDE.md]:
```suggestion
1. [`kustomize 1.x`](https://github.com/kubernetes-sigs/kustomize/releases/tag/v1.0.11) To customize YAMLs for different environments
 
```

File: docs/DEVELOPER_GUIDE.md [hunk 4]
```diff
@@ -1 +1,173 @@
...
+kubectl apply -f docs/samples/tensorflow.yaml
+```
+You should see model serving deployment running under default or your specified namespace.
+
+```console
+kubectl get pods -n default -l serving.knative.dev/configuration=my-model-default
+NAME                                                READY   STATUS    RESTARTS   AGE
+my-model-default-htz8r-deployment-8fd979f9b-w2qbv   3/3     Running   0          10s
+```
+
```
Comment (rakelkar) [inline docs/DEVELOPER_GUIDE.md]: Also add a curl to make sure ingress is setup correctly?

File: docs/DEVELOPER_GUIDE.md [hunk 5]
```diff
@@ -1 +1,173 @@
...
+- **If you change a package's deps** (including adding external dep), then you
+  must run `dep ensure`.
+
+These are both idempotent, and we expect that running these at `HEAD` to have no
+diffs. Code generation and dependencies are automatically checked to produce no
+diffs for each pull request.
+
+In some cases, if newer dependencies
+are required, you need to run "dep ensure -update package-name" manually.
```
Comment (rakelkar) [inline docs/DEVELOPER_GUIDE.md]: possibly add a comment to say that dependency changes should be a separate commit and not mixed with logic changes?

File: docs/DEVELOPER_GUIDE.md [hunk 6]
```diff
@@ -1 +1,173 @@
...
+
+## Prerequisites
+
+Follow the instructions below to set up your development environment. Once you
+meet these requirements, you can make changes and
+[deploy your own version of kfserving](#deploy-kfserving)!
+
+Before submitting a PR, see also [CONTRIBUTING.md](../CONTRIBUTING.md).
+
+### Sign up for GitHub
```
Comment (animeshsingh) [inline docs/DEVELOPER_GUIDE.md]: This might be an overkill. Having a github account is sort of expected

File: docs/DEVELOPER_GUIDE.md (hunk 1 above)
Comment (animeshsingh) [reply docs/DEVELOPER_GUIDE.md]: Essentially mention to install 'Knative Serving' only.

File: docs/DEVELOPER_GUIDE.md (hunk 3 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: +1

File: docs/DEVELOPER_GUIDE.md (hunk 5 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: +1

File: docs/DEVELOPER_GUIDE.md (hunk 2 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: Added :)

File: docs/DEVELOPER_GUIDE.md (hunk 1 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: SGTM

File: docs/DEVELOPER_GUIDE.md (hunk 4 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: I will leave this later until we have an end to end example working :)

File: docs/DEVELOPER_GUIDE.md [hunk 7]
```diff
@@ -1 +1,179 @@
-todo
\ No newline at end of file
+# Development
```
Comment (gyliu513) [inline docs/DEVELOPER_GUIDE.md]: For all of the markdown files, how about use `doctoc` to generate a toc for the document?

File: docs/DEVELOPER_GUIDE.md [hunk 8]
```diff
@@ -1 +1,179 @@
...
+1. `$GOPATH/bin` on `PATH`: This is so that tooling installed via `go get` will
+   work properly.
+1. `KO_DOCKER_REPO`: The docker repository to which developer images should be
+   pushed (e.g. `docker.io/<username>/[project]`).
+
+- **Note**: Set up a docker repository for pushing images. You can use any container image registry by adjusting 
+the authentication methods and repository paths mentioned in the sections below.
+   - [Google Container Registry quickstart](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+   - [Docker Hub quickstart](https://docs.docker.com/docker-hub/)
+   - [Azure Container Registry quickstart](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-portal)
```
Comment (gyliu513) [inline docs/DEVELOPER_GUIDE.md]: quay.io is also a good registry https://quay.io/tutorial/

File: docs/DEVELOPER_GUIDE.md [hunk 9]
```diff
@@ -1 +1,179 @@
...
+Once you reach this point you are ready to do a full build and deploy as
+described below.
+
+## Deploy KFServing
+
+### Check Knative Serving installation
+Once you've [setup your development environment](#prerequisites), you can see things running with:
+
+```console
+kubectl -n knative-serving get pods
```
Comment (gyliu513) [inline docs/DEVELOPER_GUIDE.md]: How about adding a `$` or `#` right before this command to distinguish the CLI and output? Ditto for others.

File: docs/DEVELOPER_GUIDE.md [hunk 10]
```diff
@@ -1 +1,179 @@
-todo
\ No newline at end of file
+# Development
+
+This doc explains how to setup a development environment so you can get started
+[contributing](../CONTRIBUTING.md)
+to `kfserving`. Also take a look at:
+
+- [How to add and run tests](../test/README.md)
+- [Iterating](#iterating)
```
Comment (ellis-bigelow) [inline docs/DEVELOPER_GUIDE.md]: I think this might be a bad link? Capital I?

File: docs/DEVELOPER_GUIDE.md [hunk 11]
```diff
@@ -1 +1,179 @@
...
+
+1. Create your own
+   [fork of this repo](https://help.github.com/articles/fork-a-repo/)
+1. Clone it to your machine:
+
+```shell
+mkdir -p ${GOPATH}/src/github.com/kubeflow
+cd ${GOPATH}/src/github.com/kubeflow
+git clone git@github.com:${YOUR_GITHUB_USERNAME}/kfserving.git
+cd serving
```
Comment (ellis-bigelow) [inline docs/DEVELOPER_GUIDE.md]: cd kfserving ;)

File: docs/DEVELOPER_GUIDE.md [hunk 12]
```diff
@@ -1 +1,179 @@
...
+
+
+### Smoke test after deployment
+```bash
+kubectl apply -f docs/samples/tensorflow.yaml
+```
+You should see model serving deployment running under default or your specified namespace.
+
+```console
+kubectl get pods -n default -l serving.knative.dev/configuration=my-model-default
```
Comment (ellis-bigelow) [inline docs/DEVELOPER_GUIDE.md]: We should 100% label our pods with serving.kubeflow.org/kfservice=my-model

File: docs/DEVELOPER_GUIDE.md [hunk 13]
```diff
@@ -1 +1,179 @@
...
+of:
+
+- **If you change an input to generated code**, then you must run
+  `make manifests`. Inputs include:
+
+  - API type definitions in
+    [pkg/apis/serving/v1alpha1/](../pkg/apis/serving/v1alpha1/.),
+  - Manifests or kustomize patches stored in [config](../config).
+
+- **If you change a package's deps** (including adding external dep), then you
```
Comment (ellis-bigelow) [inline docs/DEVELOPER_GUIDE.md]: To mitigate security implications of vendor attack injection, maintainers will be responsible for updating dependencies. Please cut an issue.

File: docs/DEVELOPER_GUIDE.md [hunk 14]
```diff
@@ -1 +1,179 @@
...
+diffs for each pull request.
+
+In some cases, if newer dependencies
+are required, you need to run "dep ensure -update package-name" manually.
+
+Once the codegen and dependency information is correct, redeploying the
+controller is simply:
+
+```shell
+make deploy-test
```
Comment (ellis-bigelow) [inline docs/DEVELOPER_GUIDE.md]: I think this target should be deploy-dev. wdyt?

File: docs/DEVELOPER_GUIDE.md (hunk 8 above)
Comment (ellis-bigelow) [reply docs/DEVELOPER_GUIDE.md]: Quay has had several large scale outages over the last year. My preference would be to use GCR, though perhaps we can maintain across multiple cloud vendors.

**Comment by ellis-bigelow** [comment]
Wow Dan this is awesome! Thank you!

File: docs/DEVELOPER_GUIDE.md (hunk 10 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: hmm works fine for me

File: docs/DEVELOPER_GUIDE.md (hunk 12 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: ye we can add this

File: docs/DEVELOPER_GUIDE.md (hunk 7 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: get the doctoc working looks nice :), we will need to update other docs, going to send a separate PR.

File: docs/DEVELOPER_GUIDE.md (hunk 8 above)
Comment (yuzisun) [reply docs/DEVELOPER_GUIDE.md]: this is for developer's testing image registry, three choices should be good enough :)

Comment (gyliu513) [reply docs/DEVELOPER_GUIDE.md]: For some developers in China, `quay.io` maybe a good choice. As Google Container Registry request `enable billing` and the `billing` now do not support Chinese credit card.

image

**Comment by ellis-bigelow** [comment]
Dan, feel free to approve whenever you feel comments are done.
/lgtm
